File: Majority Element/majorityElement.java
import java.util.Arrays;

class Solution {
    public int majorityElement(int[] nums) {
        /* Solution 1 */
        // multiple really good approaches to this, first being sorting and grabbing
        // the element in the middle. because we know the element will appear more than
        // n/2 times, we can sort the array and grab the middle element.
        Arrays.sort(nums);
        return nums[nums.length / 2];
    }

    public int majorityElement2(int[] nums){
        /* Solution 2 */
        // another is to use Moore's Voting algorithm to choose a candidate. builds on idea
        // that the element most present will automatically be incremented most whenever
        // it is most prevalent and will end as our choice. algorithm follows that we 
        // increment whenever we encounter current choice, and decrement otherwise
        int cnt = 0;
        int choice = 0;

        for(int i = 0; i < nums.length; i++){ // iterate list
            if(cnt == 0) // either beginning or we know the majority isn't chosen 
                choice = nums[i]; // reset choice
            
            if(nums[i] == choice) //if we encounter our choice again
                cnt++; // increment counter
            else
                cnt--; // decrement counter
        }

        return choice;
    }
}

File: Majority Element/majority_element.c
int majorityElement(int* nums, int numsSize) {
    // Sort the array in ascending order using bubble sort
    for (int i = 0; i < numsSize - 1; i++) {
        for (int j = 0; j < numsSize - i - 1; j++) {
            if (nums[j] > nums[j + 1]) { // element before is bigger than next
                int temp = nums[j]; // swap the elements
                nums[j] = nums[j + 1];
                nums[j + 1] = temp;
            }
        }
    }
    
    // Return the element at the middle index 
    return nums[numsSize / 2];
}